src/librustc_mir/transform/mod.rs

```diff
@@ -296,6 +297,10 @@ fn run_optimization_passes<'tcx>(
         // `AddRetag` needs to run after `ElaborateDrops`. Otherwise it should run fairly late,
         // but before optimizations begin.
         &add_retag::AddRetag,
+        // If the `instrument-coverage` option is enabled, analyze the CFG, identify each
+        // conditional branch, construct a coverage map to be passed to LLVM, and inject counters
+        // where needed.
+        &instrument_coverage::InstrumentCoverage,
```

Comment:
I'm now thinking that we should run this pass as early as possible before any optimization passes (i.e. before the first simplify_branches pass, _possibly_ after no_landing_pads). All the branches should look like they do in the source at that point. This will prevent some optimizations, but I think the priority should be accurate coverage info before all else.

Reply:
@tmandry - As we discussed, I moved the `instrument_coverage` pass to the end of `mir_validate()` passes.

@wesleywiser - Tyler and I tried a few places including putting the coverage pass at the top of `post_borrowck_cleanup`. We agreed that it doesn't make sense to run `instrument_coverage` more than once per function. For reasons I don't yet understand, every `println!()` seems to cause `run_optimization_passes()` to run two additional times. (With 2 `println!()`calls in `main()` for example, my debug statement showed `instrument_coverage` ran 5 times on `main()` alone.)

Let me know if you disagree with the new location.

Reply:
So I think what's actually happening is that when you add `println!()`s to the function body, some of the code added by that macro is eligible for [static promotion](https://github.com/rust-lang/rfcs/blob/master/text/1414-rvalue_static_promotion.md). Each promoted value gets its own `Body` which your code is then invoked for. Because these values were extracted out of the function, they use the function's `DefId` so when you call `debug!("instrumenting {:?}", src.def_id());`, you get the same `DefId` multiple times. If you log the full `src`, it should show that we're not running on the same `body` every time. 

You can see this by looking at the [Rust playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ffb4b7013cea287855e78c81a383deb4). For a basic "hello world" application, if you click the "Show MIR" button (in the overflow menu button on the left hand side), you'll see that there is one body for `fn main()`, one body for a `&[std::fmt::ArgumentV1; 0]` which got promoted and one body for a `&[&str; 1]` which also got promoted. 

All that said, I don't have a problem with landing this as-is. We can always tweak this later and I think it's more important that we land something so we can start iterating on the coverage info generation.